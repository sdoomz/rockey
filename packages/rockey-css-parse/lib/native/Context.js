export const PREFIX_TYPES = {
  SPACE: 1,
  DOT: 2,
  HASH: 3,
  COLUMN: 4,
  NOT: 5,
};

export default class Context {
  buffer = '';
  output = [];

  condition = null;
  selectors = {};
  css = '';

  prefix = null;
  prefixSymbol = '';

  constructor({ updateClassName } = {}) {
    this.updateClassName = updateClassName;
  }

  trim() {
    this.buffer = this.buffer.trim();
  }

  append(char) {
    this.buffer += char;
  }

  clear() {
    this.buffer = '';
  }

  appendCSS(css) {
    this.css += css;
  }

  addImport() {
    // todo
  }

  mediaStart(condition) {
    if (this.condition) {
      // TODO: merge conditions
    }

    this.condition = condition;
  }

  mediaEnd() {
    this.condition = null;
  }

  ruleStart() {
    this.selectors = {};
    this.css = '';
  }

  ruleEnd() {
    this.output.push({
      classnames: this.selectors,
      css: this.css,
    });
  }

  addProperty(prop) {
    this.appendCSS(`${prop}:`);
  }

  addValue(value) {
    this.appendCSS(`${value};`);
  }

  declarationStart() {
    this.appendCSS('{');
  }

  declarationEnd() {
    this.appendCSS('}');
  }

  addSelector(selector, type, symbol) {
    let processedSelector = selector;

    switch (type) {
      case PREFIX_TYPES.DOT:
        processedSelector = this.updateClassName
          ? this.updateClassName(selector, type, symbol)
          : selector;

        this.selectors[selector] = processedSelector;
        break;
    }

    this.css += `${symbol}${processedSelector}`;
  }

  export() {
    return this.output;
  }
}
