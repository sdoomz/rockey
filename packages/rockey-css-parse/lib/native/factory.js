import {
  rule,
  branch,
  label,
  link,
  propertyValue,
  property,
  step,
  block,
  selector,
  alphabetical,
  Schema,
} from '../schema';

import { ALPHABETICAL } from '../schema/markers';

const COMMENTS_BLOCK = 'COMMENTS';
const SELECTOR_BLOCK = 'SELECTOR';
const MEDIA_BLOCK = 'MEDIA';
const IMPORT_BLOCK = 'IMPORT';

const RULE_BLOCK = 'RULE';

function clearContext(context) {
  context.clear();
}

export default function factory({
  actions: {
    // selectors
    startTagSelector,
    startClassSelector,
    startIdSelector,
    startPseudoSelector,
    //
    selectorEnd,
    selectorComaSeparated,
    selectorSpaceSeparated,
    pseudoSelector,
    withParentSelector,
    // declaraion
    declarationStart,
    declarationEnd,
    // prop
    propertyStart,
    propertyEnd,
    // value
    valueStart,
    valueEnd,

    //---
    startChildWithPseudo,
    endChildWithPseudo,
  },
}) {
  const schema = new Schema();

  schema.block(
    IMPORT_BLOCK,
    step('@import'),
    step(';', context => {
      context.appendRule(context.buffer);
    })
  );

  schema.block(
    MEDIA_BLOCK,
    {
      onEnter: context => {
        context.startMedia();
      },
      onLeave: context => {
        context.endMedia();
      },
    },
    step('@media', clearContext, context => {
      context.addCondition(context.buffer.trim());
    }),
    step('{', clearContext),
    block(RULE_BLOCK),
    step('}', {
      endAction: clearContext,
    }),
    block.out()
  );

  schema.block(
    COMMENTS_BLOCK,
    step('/*'),
    step('*/', {
      endAction: clearContext,
    }),
    block.out()
  );

  const START_SELECTOR_LABEL = 'START_SELECTOR';
  schema.block(
    SELECTOR_BLOCK,
    // {
    //   onEnter: context => {
    //     context.startSelectors();
    //   },
    //   onLeave: context => {
    //     context.endSelectors();
    //   },
    // },
    // ----
    label(START_SELECTOR_LABEL),
    rule.repeat(
      branch(
        selector(ALPHABETICAL, startTagSelector, selectorEnd),
        selector('.', startClassSelector, selectorEnd),
        // selector('#', startIdSelector, selectorEnd),
        rule(step(':', startPseudoSelector), link(START_SELECTOR_LABEL))
      ),
      branch.maybe(
        step(' ', selectorSpaceSeparated),
        rule(step(',', selectorComaSeparated), link(START_SELECTOR_LABEL)),
        rule(step(':', pseudoSelector), link(START_SELECTOR_LABEL))
        // rule(
        //   step(':not(', context => {
        //     context.clear();
        //     context.startNotSelector();
        //   }),
        //   block(SELECTOR_BLOCK),
        //   step(')', context => {
        //     context.clear();
        //     context.endNotSelector();
        //   })
        // )
      )
    ),

    block.out()
  );

  schema.block(
    RULE_BLOCK,
    {
      onEnter: context => {
        context.startRule();
      },
      onLeave: context => {
        context.endRule();
      },
    },

    block(SELECTOR_BLOCK),
    step('{', declarationStart),

    rule.repeat.maybe(
      branch(
        block(COMMENTS_BLOCK),
        rule(
          property(propertyStart, propertyEnd),
          step(':', clearContext),
          propertyValue(valueStart, valueEnd),
          step(';', clearContext)
        )
      )
    ),

    step('}', declarationEnd),
    block.out()
  );

  // prettier-ignore
  schema.main(
    rule.repeat(
      block(RULE_BLOCK), 
      // branch(
        
      //   block(COMMENTS_BLOCK),
      //   block(MEDIA_BLOCK)
      // )
    )
  );

  return schema;
}
