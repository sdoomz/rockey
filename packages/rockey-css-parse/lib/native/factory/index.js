import { rule, branch, block, Schema } from 'rockster/schema';

import registerSelectors, {
  blocks as SELECTORS_BLOCKS,
} from './blocks/selectors';
import registerMedia, { blocks as MEDIA_BLOCKS } from './blocks/media';
import registerComments, { blocks as COMMENTS_BLOCKS } from './blocks/comments';
import registerImport, { blocks as IMPORT_BLOCKS } from './blocks/import';
import registerCSSRule, { blocks as CSS_RULE_BLOCKS } from './blocks/rule';

export default function factory() {
  const schema = new Schema();

  const blocks = {
    ...SELECTORS_BLOCKS,
    ...MEDIA_BLOCKS,
    ...IMPORT_BLOCKS,
    ...COMMENTS_BLOCKS,
    ...CSS_RULE_BLOCKS,
  };

  registerSelectors({
    schema,
    blocks,
  });

  registerMedia({
    blocks,
    schema,
  });

  registerComments({
    blocks,
    schema,
  });

  registerImport({
    blocks,
    schema,
  });

  registerCSSRule({
    blocks,
    schema,
  });

  schema.main(
    rule.repeat(
      // branch(
      block(CSS_RULE_BLOCK)
      // block(MEDIA_BLOCK),
      // block(COMMENTS_BLOCK),
      // block(IMPORT_BLOCK)
      // )
    )
  );

  return schema;
}
