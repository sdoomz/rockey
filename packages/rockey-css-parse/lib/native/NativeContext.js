import Context from 'rockster/Context';
import { selector } from '../schema/index';

export const SELECTOR_TYPES = {
  CLASS: 0,
  ID: 1,
  PSEUDO: 2,
  TAG: 3,
};

export const SELECTOR_SEPARATORS = {
  NONE: 0,
  COMMA: 1,
  WITH_PARENT: 2,
  SPACE: 3,
  PSEUDO: 4,
  NOT: 5,
};

const NESTED_SELECTORS = {
  NONE: 0,
  NOT_SELECTOR: 1,
};

export default class CSSContext extends Context {
  selectorSeparator = SELECTOR_SEPARATORS.NONE;

  condition = '';

  css = '';
  output = [];

  property = '';

  parentSelectors = '';

  hasProperties = false;

  openedNotSelectors = 0;
  // nestedSelectorType = NESTED_SELECTORS.NONE;
  // selectors = '';
  // selectorsMap = {};
  // selectorsBackups = [];

  classnames = {};

  constructor(options = {}) {
    super(options);
    this.updateSelector = options.updateSelector;
  }

  startRule() {
    this.hasProperties = false;
    this.classnames = {};
    this.css = '';
  }

  endRule() {
    if (this.hasProperties) {
      if (this.condition) {
        this.output.push(`@media${this.condition}{${this.css}}`);
        // this.output.push({
        //   selectors: {},
        //   css: `@media${this.condition}{${this.css}}`,
        // });
      } else {
        // this.output.push(this.css);
        this.output.push({
          classnames: this.classnames,
          css: `${this.css}`,
        });
      }
    }
  }

  startMedia() {
    this.condition = '';
  }

  endMedia() {
    this.condition = null;
    this.appendCSS('}');
  }

  addCondition(condition) {
    this.condition += condition;
  }

  appendCSS(css) {
    this.css += css;
  }

  appendRule(rule) {
    this.output.push(rule);
  }

  startNotSelector() {
    this.appendCSS(':not(');
    this.selectorSeparator = SELECTOR_SEPARATORS.NONE;
    // this.openedNotSelectors++;
    // this.nestedSelectorType = NESTED_SELECTORS.NOT_SELECTOR;
  }

  endNotSelector() {
    // this.openedNotSelectors--;
    this.appendCSS(')');
    this.selectorSeparator = SELECTOR_SEPARATORS.NONE;
    // this.nestedSelectorType = NESTED_SELECTORS.NONE;
  }

  startSelectors() {
    // this.selectorsBackups.push({
    //   selectors: this.selectors,
    //   selectorSeparator: this.selectorSeparator,
    //   nestedSelectorType: this.nestedSelectorType,
    // });
    // this.selectorSeparator = SELECTOR_SEPARATORS.NONE;
    // this.selectors = '';
  }

  endSelectors() {
    // const selectors = this.selectors;
    // if (this.selectorsBackups.length) {
    //   const memo = this.selectorsBackups.pop();
    //   this.selectorSeparator = memo.selectorSeparator;
    //   this.selectors = memo.selectors;
    //   this.nestedSelectorType = memo.nestedSelectorType;
    // } else {
    //   this.selectors = '';
    //   xx;
    // }
    // console.log(selectors);
    // switch (this.nestedSelectorType) {
    //   case NESTED_SELECTORS.NONE:
    //     this.selectors = `${this.selectors}${selectors}`;
    //     break;
    //   case NESTED_SELECTORS.NOT_SELECTOR:
    //     this.selectors = `${this.selectors}:not(${selectors})`;
    //     break;
    // }
    // console.log(this.selectors);
  }

  addSelector(raw) {
    let selector = this.updateSelector
      ? this.updateSelector(raw, this.selectorType)
      : raw;

    switch (this.selectorType) {
      case SELECTOR_TYPES.CLASS:
        this.classnames[raw.slice(1)] = selector;
        break;
    }

    let appendSelector = selector;

    switch (this.selectorSeparator) {
      // case SELECTOR_SEPARATORS.NOT:
      //   this.appendCSS(`:not(${appendSelector}`);
      //   break;

      case SELECTOR_SEPARATORS.COMMA:
        this.appendCSS(`,${appendSelector}`);

        break;
      case SELECTOR_SEPARATORS.SPACE:
        this.appendCSS(` ${appendSelector}`);

        break;
      case SELECTOR_SEPARATORS.PSEUDO:
      case SELECTOR_SEPARATORS.NONE:
        this.appendCSS(appendSelector);

        break;
    }
  }

  export() {
    return this.output;
  }
}
