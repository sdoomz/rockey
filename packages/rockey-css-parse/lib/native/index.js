import prepare from 'rockster/parsers/crossrefs/encode';
import createParser from 'rockster/parsers/crossrefs';

import factory from './factory/index';
import Context from './Context';

const schema = factory();

const segments = prepare(schema.compose());

export default function create(options) {
  return createParser({
    createContext: () => new Context(options),
    segments,
  });
}
