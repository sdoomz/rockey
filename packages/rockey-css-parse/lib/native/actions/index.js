import { SELECTOR_SEPARATORS, SELECTOR_TYPES } from '../NativeContext';

export function startTagSelector(context) {
  context.selectorType = SELECTOR_TYPES.TAG;
}

export function startClassSelector(context) {
  context.selectorType = SELECTOR_TYPES.CLASS;
}

export function startIdSelector(context) {
  context.selectorType = SELECTOR_TYPES.ID;
}

export function startPseudoSelector(context) {
  context.selectorType = SELECTOR_TYPES.PSEUDO;
}

export function selectorEnd(context) {
  context.addSelector(context.buffer.trim());
  context.clear();
}

export function selectorComaSeparated(context) {
  context.selectorSeparator = SELECTOR_SEPARATORS.COMMA;
  context.clear();
}

export const pseudoSelector = context => {
  context.selectorSeparator = SELECTOR_SEPARATORS.PSEUDO;
};

export const withParentSelector = context => {
  context.clear();
  context.selectorSeparator = SELECTOR_SEPARATORS.WITH_PARENT;
};

export const selectorSpaceSeparated = context => {
  context.selectorSeparator = SELECTOR_SEPARATORS.SPACE;
  context.clear();
};

// declaraion
export const declarationStart = context => {
  context.selectorSeparator = SELECTOR_SEPARATORS.NONE;
  context.clear();
  context.appendCSS('{');
};

export const declarationEnd = context => {
  context.appendCSS('}');
  context.clear();
};

// prop
export const propertyStart = context => {};

export const propertyEnd = context => {
  context.property = context.buffer.trim();
  context.clear();
};

// value
export const valueStart = context => {};

export const valueEnd = context => {
  context.trim();
  context.hasProperties = true;

  context.appendCSS(`${context.property}:${context.buffer};`);
  context.clear();
  context.property = '';
};

export const startChildWithPseudo = context => {};
export const endChildWithPseudo = context => {};
