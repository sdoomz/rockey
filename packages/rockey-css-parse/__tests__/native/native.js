import create from '../../native';
import { SELECTOR_TYPES } from '../../native/NativeContext';

let id = 0;

function resetId() {
  id = 0;
}

const parse = create({
  updateSelector: (selector, type) => {
    switch (type) {
      case SELECTOR_TYPES.CLASS:
      case SELECTOR_TYPES.ID:
        return `${selector}-${++id}`;
      default:
        return selector;
    }
  },
});

describe('native css', () => {
  beforeEach(() => resetId());

  test('simple', () => {
    const context = parse(`
.bar { 
  color: yellow;
}`);

    expect(context.export()).toMatchSnapshot();
  });

  test('few rules', () => {
    const context = parse(`
.bar { 
  color: yellow;
}

.foo {
  color: orange; 
}`);

    expect(context.export()).toMatchSnapshot();
  });
});
