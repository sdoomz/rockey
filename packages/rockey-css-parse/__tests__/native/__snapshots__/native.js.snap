// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`native css few rules 1`] = `
Array [
  Object {
    "classnames": Object {
      "bar": ".bar-1",
    },
    "css": ".bar-1{color:yellow;}",
  },
  Object {
    "classnames": Object {
      "foo": ".foo-2",
    },
    "css": ".foo-2{color:orange;}",
  },
]
`;

exports[`native css simple 1`] = `
Array [
  Object {
    "classnames": Object {
      "bar": ".bar-1",
    },
    "css": ".bar-1{color:yellow;}",
  },
]
`;
